data_structures/binary_search_tree.ts

```diff
@@ -0,0 +1,199 @@
+/**
+ * Represents a node of a binary search tree.
+ *
+ * @template T The type of the value stored in the node.
+ */
+class TreeNode<T> {
+  constructor(
+    public data: T,
+    public leftChild?: TreeNode<T>,
+    public rightChild?: TreeNode<T>,
+  ) {}
+}
+
+/**
+ * An implementation of a binary search tree.
+ *
+ * A binary tree is a tree with only two children per node. A binary search tree on top sorts the children according
+ * to following rules:
+ * - left child < parent node
+ * - right child > parent node
+ * - all children on the left side < root node
+ * - all children on the right side > root node
+ *
+ * For profound information about trees
+ * @see https://www.geeksforgeeks.org/introduction-to-tree-data-structure-and-algorithm-tutorials/
+ *
+ * @template T The data type of the values in the binary tree.
+ */
+export class BinarySearchTree<T> {
+  rootNode?: TreeNode<T>;
+
+  /**
+   * Instantiates the binary search tree.
+   *
+   * @param rootNode The root node.
+   */
+  constructor() {
+    this.rootNode = undefined;
+  }
+
+  /**
+   * Checks, if the binary search tree is empty, i. e. has no root node.
+   *
+   * @returns Whether the binary search tree is empty.
+   */
+  isEmpty(): boolean {
+    return this.rootNode === undefined;
+  }
+
+  /**
+   * Searches for data in the current binary search tree.
+   *
+   * @param data The data to search for.
+   */
+  search(data: T): TreeNode<T> {
```

Comment:
`search` shouldn't return a `TreeNode`, but rather the datum `T` that was found

Reply:
I changed it, but I'd like to know why exactly I should return the datum `T`. I assume it's because `TreeNode<T>` should not be exposed, but returning the same data you give as parameter to the function confuses me a bit. :sweat_smile:

Reply:
Yes. It is completely reasonable to implement a `has` instead of a `find` function, and at this stage, it probably is preferable.

The reason for a `search` function returning the datum is that you usually have *sentinel data* attached to the *key*. You can implement a sorted map in terms of a sorted set by inserting entries `{key: key, value: value}` and having a comparator that compares keys. To be able to get out the value, you need the `search` function to return the found entry, which consists of key & value.

That said, since you haven't parameterized this in terms of a custom comparator yet, it is indeed preferable to simply change this to a `has` method which returns a boolean.

Reply:
Seems logical, I have now changed the search function to a has function.